algorithms/src/Kokkos_RadixSort.hpp

```diff
@@ -0,0 +1,251 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_RADIXSORT_HPP_
+#define KOKKOS_RADIXSORT_HPP_
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+#define KOKKOS_IMPL_PUBLIC_INCLUDE
+#define KOKKOS_IMPL_PUBLIC_INCLUDE_NOTDEFINED_RADIXSORT
+#endif
+
+#include <Kokkos_Core.hpp>
+
+namespace Kokkos {
+namespace Experimental {
+
+// https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda
+
+template <typename KeyView>
+struct KeyFromView {
+  using key_value_type                    = typename KeyView::value_type;
+  static constexpr std::uint32_t num_bits = sizeof(key_value_type) * 8;
+
+  KeyView keys;
+  KeyFromView(KeyView k) : keys(k) {}
+
+  // i: index of the key to get
+  // bit: which bit, with 0 indicating the least-significant
+  auto operator()(int i, std::uint32_t bit) const {
+    auto h = keys(i) >> bit;
+
+    // Handle signed 2's-complement
+    if constexpr (std::is_signed_v<key_value_type>) {
+      if (bit == num_bits - 1) {
+        h = ~h;
+      }
+    }
+    return ~h & 0x1;
+  }
+};
+
+template <typename T, typename IndexType = ::std::uint32_t>
+class RadixSorter {
+ public:
+  static_assert(std::is_integral_v<T>, "Keys must be integral for now");
+
+  static constexpr std::uint32_t num_bits = sizeof(T) * 8;
+
+  RadixSorter() = default;
+  explicit RadixSorter(std::size_t n)
+      : m_key_scratch("radix_sort_key_scratch", n),
+        m_index_old("radix_sort_index", n),
+        m_index_new("radix_sort_index_scratch", n),
+        m_scan("radix_sort_scan", n),
+        m_bits("radix_sort_bits", n) {}
+
+  // Generate and store the permutation induced by the keys, without
+  // modifying their initial order
+  template <class ExecutionSpace>
+  void create_indirection_vector(ExecutionSpace const& exec, View<T*> keys) {
+    auto key_functor = KeyFromView{keys};
+    const auto n     = keys.extent(0);
+
+    create_indirection_vector(exec, key_functor, n);
+  }
+
+  template <class ExecutionSpace, class KeyFunctor>
+  void create_indirection_vector(ExecutionSpace const& exec,
+                                 KeyFunctor key_functor, size_t n) {
+    RangePolicy<ExecutionSpace> policy(exec, 0, n);
+
+    // Initialize m_index_old, since it will be read from in the first
+    // iteration's call to step()
+    Kokkos::parallel_for(
+        policy, KOKKOS_LAMBDA(int i) { m_index_old(i) = i; });
+
+    for (int i = 0; i < num_bits; ++i) {
```

Comment:
These comparisons will need to change to the same signed-ness between `i` and `num_bits`

Reply:
Done